test/EcdsaMultikey.spec.js

```diff
@@ -173,7 +171,7 @@ describe('EcdsaMultikey', () => {
 
 function _ensurePublicKeyEncoding({keyPair, publicKeyMultibase}) {
   keyPair.publicKeyMultibase.startsWith(MULTIBASE_BASE58_HEADER).should.be.true;
-  const decodedPubkey = baseDecode(publicKeyMultibase);
-  const encodedPubkey = baseEncode(decodedPubkey);
+  const decodedPubkey = base58.decode(publicKeyMultibase);
+  const encodedPubkey = base58.encode(decodedPubkey);
```

Comment:
So are `base58.decode` and `base58.encode` really drop in replacements for `baseDecode` and `baseEncode`? I would think the multiformats versions check for and include the leading `z` character whereas the `base58.*` versions do not. If that's the case, it's also clear that the tests need a little tweaking to ensure that this difference gets caught somehow.

So I'd expect us to need to check for `z` and add `z`.

Reply:
After testing, it turns out that the encode and decode functions from `base58-universal` behave the same as those from the `multiformats` library.

Reply:
In my own testing, the contents of the `Uint8Array` that results from `multibase.decode(publicKeyMultibase)` does not equal the contents of the `Uint8Array` from `base58.decode(publicKeyMultibase)`. However, it does equal the contents of the `Uint8Array` from `base58.decode(publicKeyMultibase.slice(1))`, confirming that `base58-universal` does not consider the multibase prefix `z` that indicates a base58 encoding.

If you check the source code for `base58-universal` you can also confirm this; that library only does the base58 encoding/decoding, it does not know anything about multibase.

Can you please recheck or indicate how you were performing the test you mentioned?

Reply:
Interesting - I am still seeing the same results between `base58-universal@2.0.0` and `multiformats@11.0.1` from my testing. The value of `publicKeyMultibase` that I am using is the same one from `test/mock-data.js` (`zDnaeSMnptAKpH4AD41vTkwzjznW7yNetdRh9FJn8bJsbsdbw`). And the code I am using to test is as follows:

```
import * as base58 from 'base58-universal';
import {base58btc} from 'multiformats/bases/base58';
const {baseDecode, baseEncode} = base58btc;

const decodedPubkeyB58 = base58.decode(publicKeyMultibase);
const encodedPubkeyB58 = base58.encode(decodedPubkeyB58);
const decodedPubkeyMF = baseDecode(publicKeyMultibase);
const encodedPubkeyMF = baseEncode(decodedPubkeyMF);

console.log('decoded pub key b58:', decodedPubkeyB58);
console.log('encoded pub key b58:', encodedPubkeyB58);
console.log('decoded pub key mf:', decodedPubkeyMF);
console.log('encoded pub key mf:', encodedPubkeyMF);
```

Reply:
@kezike,

The previous code that used the `multibase` library (`multibase.decode`) was doing the proper transformation -- and I thought that the new `baseDecode` function used here was doing the same. It's not; it's just doing base58-decoding, so using it would also be the wrong thing to do. That also means that just substituting it with `base58.decode` is also the wrong thing to do.

So, it's true that `baseDecode` and `base58.decode` (and their encoding counterparts) perform the same function. The problem is that the first character in `publicKeyMultibase` is not part of the base58-encoded value, it's a multibase header. But the above test is erroneously treating like it's part of the encoded value. The letter `z` has a special meaning that means "everything after `z` is base58-encoded".

The code needs to make sure it does not mistakenly treat the `z` like it's just another character in the base58-encoded value, as if it's just another symbol in the base58 alphabet that represents part of the multicodec-encoded public key. Instead, the first character must be stripped off before decoding (and it must be checked to make sure it's a `z`). Only then can the multicodec-encoded public key be decoded into a byte array. When encoding, the `z` header will need to be added back.

For clarity, this is how the encoding works:

1. Start with `<raw public key bytes>`.
2. Add a multicodec header that describes that the bytes are for a public key and one of a specific type: `<multicodec header><raw public key bytes>`. This forms the multicodec / multikey representation of the public key.
3. Now, to represent that as a string, encode it using the base58 alphabet: `base58Encode(<multicodec header><raw public key bytes>)`.
4. Now, finally, add ***another header*** that says which alphabet was used. In multibase, the letter `z` means the base58 alphabet was used. As a counterexample, the letter `u` says that the base64url alphabet was used. If we had encoded using base64url ... our final string would start with `u`. That character *is not* part of the base-encoded value, however, it is part of the *multibase* value -- it's a header. So, since we used the base58 alphabet, we use the `z` header: `z` + `base58Encode(<multicodec header><raw public key bytes>)`.

You can see all of the registered multibase headers here:
https://github.com/multiformats/multibase/blob/master/multibase.csv

Hopefully this fully explains what's going on. The error is that the multibase header (`z` in this case) is being treated as part of the base-encoded value... when it's not.